File: java/Glacier2/session/server/src/main/java/com/example/glacier2/session/server/InMemoryPokeStore.java
// Copyright (c) ZeroC, Inc.

package com.example.glacier2.session.server;

import java.util.ArrayList;
import java.util.Hap;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class InMemoryPokeStore implements PokeStore {
    private final Map<String, List<String>> _store = new HashMap<>();

    public void saveCollection(String userId, List<String> pokemon) {
        if (pokemon.isEmpty()) {
            _store.remove(userId);
        } else {
            _store.put(userId, List.copyOf(pokemon)); // 'copyOf' returns an immutable list.
        }
    }

    public List<String> retrieveCollection(String userId) {
        return _store.getOrDefault(userId, null);
    }
}

File: cpp/Glacier2/session/InMemoryPokeStore.cpp

#include "InMemoryPokeStore.h"

using namespace Server;
using namespace std;

void
InMemoryPokeStore::saveCollection(const string& userId, CatchThemAll::PokemonList pokemon)
{
    if (pokemon.empty())
    {
        _store.erase(userId);
    }
    else
    {
        _store[userId] = std::move(pokemon);
    }
}

CatchThemAll::PokemonList
InMemoryPokeStore::retrieveCollection(const string& userId) const
{
    auto it = _store.find(userId);
    return it == _store.end() ? CatchThemAll::PokemonList{} : it->second;
}
